### PR TITLE
[RFC] o/devicestate: don't prune tasks immediediately

### PR DESCRIPTION
**Pull request by stolowski**
I think that preseeded image may potentially have an issue with task/change pruning on first boot, because the "seed" change in the state (that's is partially done, and partially carried over on first boot) has tasks with spawn times in the past, and if more than 24hrs passed since spawn time, the change may get aborted. There is a 10-minute window defined before pruneTicker will kick in and Prune(), so theoretically it wouldn't happen, but I think it's better to explicitly prevent it until 'seeded' is true.

### CODE REVIEW

File: overlord/devicestate/devicemgr.go [hunk 1]
```diff
@@ -146,6 +146,28 @@ func (m *DeviceManager) CanStandby() bool {
...
 
+var canPruneDelay = 1 * time.Hour
+
+func (m *DeviceManager) CanPrune() bool {
+	var seeded bool
+	if err := m.state.Get("seeded", &seeded); err != nil || !seeded {
+		return false
+	}
+
+	// we are seeded. check if "seed" change is still
```
Comment (zyga) [inline overlord/devicestate/devicemgr.go]:
```suggestion
	// We are seeded. Check if "seed" change is still
```

File: overlord/devicestate/devicemgr.go [hunk 2]
```diff
@@ -146,6 +146,28 @@ func (m *DeviceManager) CanStandby() bool {
...
+	// in state and at least canPruneDelay passed since it has became ready.
+	for _, chg := range m.state.Changes() {
+		// block pruning for 1h from seeding
+		if chg.IsReady() && chg.Kind() == "seed" {
+			return time.Now().After(chg.ReadyTime().Add(canPruneDelay))
+		}
+	}
+
+	// seed change not found, so it was pruned already, don't
```
Comment (zyga) [inline overlord/devicestate/devicemgr.go]:
```suggestion
	// Seed change not found, so it was pruned already, don't
```

File: overlord/devicestate/devicemgr.go [hunk 3]
```diff
@@ -146,6 +146,28 @@ func (m *DeviceManager) CanStandby() bool {
...
+func (m *DeviceManager) CanPrune() bool {
+	var seeded bool
+	if err := m.state.Get("seeded", &seeded); err != nil || !seeded {
+		return false
+	}
+
+	// we are seeded. check if "seed" change is still
+	// in state and at least canPruneDelay passed since it has became ready.
+	for _, chg := range m.state.Changes() {
+		// block pruning for 1h from seeding
```
Comment (zyga) [inline overlord/devicestate/devicemgr.go]:
```suggestion
		// Don't prune until one hour elapses after seeding.
```

**Comment by stolowski** [comment]
I just had 'eureka' moment, I think I overcooked it and 'seeded' flag is enough.

File: overlord/overlord.go [hunk 4]
```diff
@@ -403,7 +403,9 @@ func (o *Overlord) Loop() {
 			case <-o.pruneTicker.C:
 				st := o.State()
 				st.Lock()
-				st.Prune(pruneWait, abortWait, pruneMaxChanges)
+				if o.deviceMgr == nil || o.deviceMgr.CanPrune() {
```
Comment (zyga) [inline overlord/overlord.go]: Is this an unrelated fix?

File: overlord/overlord_test.go [hunk 5]
```diff
@@ -618,6 +618,95 @@ func (ovs *overlordSuite) TestEnsureLoopPruneRunsMultipleTimes(c *C) {
...
+	c.Check(st.Changes(), HasLen, 2)
+	st.Unlock()
+
+	// setting time to currentTime, affects "now" time in Prune().
+	state.MockTime(currentTime)
+
+	// start the loop that runs the prune ticker
+	o.Loop()
+
+	time.Sleep(1500 * time.Millisecond)
```
Comment (zyga) [inline overlord/overlord_test.go]: Is there a way to avoid a 1.5s sleep in tests? Could we do it so that we get a notification that the relevant change happened?

File: overlord/overlord.go (hunk 4 above)
Comment (stolowski) [reply overlord/overlord.go]: Added a comment.

File: overlord/overlord.go [hunk 6]
```diff
@@ -403,7 +403,11 @@ func (o *Overlord) Loop() {
 			case <-o.pruneTicker.C:
 				st := o.State()
 				st.Lock()
-				st.Prune(pruneWait, abortWait, pruneMaxChanges)
+				// account for deviceMgr == nil as it's not always present in
+				// tests.
+				if o.deviceMgr == nil || o.deviceMgr.CanPrune() {
+					st.Prune(pruneWait, abortWait, pruneMaxChanges)
```
Comment (stolowski) [inline overlord/overlord.go]: Note, this change, in case of multiple failures on seeding, will result in growing state (failed "seed" changes will not get pruned until one finally succeeds, which may never happen because if it fails, it probably fails for good), I wonder if this is a valid concern. I may need to rethink this. Perhaps a better fix is inside state.Purge(), where the logic around max number of changes could always be executed. CC @pedronis

File: overlord/overlord_test.go (hunk 5 above)
Comment (stolowski) [reply overlord/overlord_test.go]: The only way to avoid this would be to call state.Prune() directly, but in all the tests in this package we want to do this indirectly via Loop which uses the mocked prune interval to be as close to real thing as possible.

**Comment by zyga** [comment]
I've shared a version that uses epoll with @mvo5 and tested that it works fine on 14.04 with golang available there. We can discuss this more tomorrow.

**Comment by pedronis** [comment]
On Mon, Feb 10, 2020, 20:52 Zygmunt Krynicki <notifications@github.com>
wrote:

> I've shared a version that uses epoll with @mvo5 <https://github.com/mvo5>
> and tested that it works fine on 14.04 with golang available there. We can
> discuss this more tomorrow.
>
>

Is this meant for Pawel's PR?

**Comment by pedronis** [comment]
@stolowski while I think this is kind of ok in practice,  it's basically saying don't prune or abort anything until we are seeded which conceptually seems a big change from normal expectations. As you mentioned there's probably a way to turn this into only affecting abort.  

But the issue we have is that the spawn time of some changes is not meaningful to drive pruning/aborting,  I think there should be a way to attack the problem in its own terms. Let's chat when we have a moment.

**Comment by stolowski** [comment]
Blocking as the approach will be discussed and changed.

**Comment by stolowski** [comment]
Closing, will be replaced by new PRs.
